core/common/src/DisplayStyleSettings.ts

```diff
@@ -1364,6 +1367,54 @@ export class DisplayStyle3dSettings extends DisplayStyleSettings {
     this._json3d.lights = lights.toJSON();
   }
 
+  /** Adjust the solar light direction based on a date and time at a geographic location.
+   * This replaces `this.lights` with a copy that records the time point and the computed direction.
+   * @param timePoint The time in UNIX milliseconds.
+   * @param location The geographic location, or an iModel from which the geolocation is to be obtained.
+   * @see [[sunTime]] to get the current sun time.
+   * @see [[clearSunTime]] to clear the time point.
+   * @note If the iModel is not geolocated, a location in Exton, Pennsylvania will be used to compute the light direction.
+   */
+  public setSunTime(timePoint: number, location: IModel | Cartographic): void {
+    let cartoCenter;
+    if (location instanceof IModel) {
+      if (location.isGeoLocated) {
+        const projectExtents = location.projectExtents;
+        const projectCenter = Point3d.createAdd2Scaled(projectExtents.low, .5, projectExtents.high, .5);

```

Comment:
@RBBentley  wouldn't it be better to use the ecef origin?

Reply:
Ask @RBBentley - I just moved his code.

Reply:
No - this is fine.   We don't really store a seperate ECEF origin so project center is best we can do.

Reply:
we do store the ecef center - that's what the ecef transform means. If the iModel is projected, we calculate the ecef center using this algorithm (though i thinlk we adjust z=0). If the iModel is not projected, the ecef origin comes from the application.

Reply:
changed to use ecef location.